### PR TITLE
[SPARK-23809][SQL] Active SparkSession should be set by getOrCreate

### PR DESCRIPTION
**Pull request by ericl**
## What changes were proposed in this pull request?

Currently, the active spark session is set inconsistently (e.g., in createDataFrame, prior to query execution). Many places in spark also incorrectly query active session when they should be calling activeSession.getOrElse(defaultSession) and so might get None even if a Spark session exists.

The semantics here can be cleaned up if we also set the active session when the default session is set.

Related: https://github.com/apache/spark/pull/20926/files

## How was this patch tested?

Unit test, existing test. Note that if https://github.com/apache/spark/pull/20926 merges first we should also update the tests there.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/test/TestSQLContext.scala
Comment (ericl) [inline sql/core/src/test/scala/org/apache/spark/sql/test/TestSQLContext.scala]: Intentional conflict with https://github.com/apache/spark/pull/20926 , should resolve once one is merged

**Comment by ericl** [comment]
@gatorsmile @marmbrus

File: sql/core/src/test/scala/org/apache/spark/sql/test/TestSQLContext.scala
Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/test/TestSQLContext.scala]: We will merge that one first.

**Comment by gatorsmile** [comment]
LGTM

**Review by felixcheung — APPROVED** [review]

**Comment by ericl** [comment]
Per @marmbrus 's suggestion also added a `SparkSession.active` call to cover any edge case where the thread-local might not be set correctly. Spark code should prefer to use this instead of getActive.get moving forward.

**Comment by gatorsmile** [comment]
@ericl Address the conflict?

**Comment by ericl** [comment]
@gatorsmile done!

**Comment by gatorsmile** [comment]
LGTM pending Jenkins

**Comment by gatorsmile** [comment]
Thanks! Merged to master

File: sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala]: @ericl Could you submit a PR to 4.x without this new API?

**Pull request closed without merging**
